File: apps/web/src/features/workflows/types/runtime.ts
import type { WorkflowNodeKind } from "@/features/workflows/types/workflow";

export type RuntimeNodeStatus =
  | "idle"
  | "queued"
  | "running"
  | "waiting"
  | "success"
  | "failed"
  | "retrying"
  | "skipped"
  | "cancelled";

export type WorkflowRunStatus =
  | "idle"
  | "running"
  | "paused"
  | "completed"
  | "stopped";

export type PlaybackSpeed = 0.5 | 1 | 2 | 5;

export type RuntimeEventType =
  | "run.started"
  | "run.paused"
  | "run.resumed"
  | "run.stopped"
  | "run.completed"
  | "node.queued"
  | "node.started"
  | "node.waiting"
  | "node.succeeded"
  | "node.failed"
  | "node.retrying"
  | "node.skipped"
  | "node.cancelled"
  | "blockchain.started"
  | "blockchain.submitted"
  | "blockchain.finalized"
  | "blockchain.failed";

export type RuntimeLogLevel = "info" | "success" | "warning" | "error";

export interface RuntimeEvent {
  id: string;
  type: RuntimeEventType;
  timestamp: number;
  nodeId?: string;
  agent?: string;
  action: string;
  status: RuntimeNodeStatus | WorkflowRunStatus;
  durationMs?: number;
  outputSummary?: string;
  level: RuntimeLogLevel;
  payload?: Record<string, unknown>;
}

export interface RuntimeNodeRecord {
  nodeId: string;
  kind: WorkflowNodeKind;
  label: string;
  status: RuntimeNodeStatus;
  input: string;
  output: string;
  outputSummary: string;
  logs: string[];
  dependencies: string[];
  durationMs: number;
  estimatedDurationMs: number;
  cost: number;
  memoryMb: number;
  retries: number;
  startedAt?: number;
  completedAt?: number;
}

export interface RuntimeMetrics {
  elapsedMs: number;
  completedNodes: number;
  totalNodes: number;
  currentAgent: string;
  memoryMb: number;
  estimatedCost: number;
  progress: number;
  successRate: number;
  activeThreads: number;
  remainingMs: number;
}

export interface WorkflowRuntimeState {
  runId: string | null;
  status: WorkflowRunStatus;
  speed: PlaybackSpeed;
  randomFailures: boolean;
  currentNodeId: string | null;
  currentIndex: number;
  orderedNodeIds: string[];
  startedAt: number | null;
  pausedAt: number | null;
  elapsedBeforePauseMs: number;
  records: Record<string, RuntimeNodeRecord>;
  events: RuntimeEvent[];
}

File: apps/api/src/agenthub/schemas/runtime.py
from datetime import datetime
from decimal import Decimal
from typing import TYPE_CHECKING, Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

if TYPE_CHECKING:
    from agenthub.models import ExecutionEvent

RuntimeStatus = Literal[
    "idle",
    "queued",
    "running",
    "waiting",
    "success",
    "failed",
    "retrying",
    "skipped",
    "cancelled",
]
RunStatus = Literal["queued", "running", "paused", "completed", "stopped", "failed"]


class WorkflowRunCreate(BaseModel):
    workflow_id: UUID
    goal: str = Field(default="", max_length=2000)
    language: str = Field(default="English", max_length=80)
    speed: Annotated[float, Field(gt=0, le=5)] = 1
    random_failures: bool = False


class InlineWorkflowRunCreate(BaseModel):
    workflow: object
    speed: Annotated[float, Field(gt=0, le=5)] = 1
    random_failures: bool = False


class WorkflowRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    workflow_id: UUID
    status: str
    speed: Decimal
    random_failures: bool
    goal: str
    language: str
    current_node_key: str | None
    progress: int
    started_at: datetime | None
    completed_at: datetime | None
    error: str | None
    created_at: datetime
    updated_at: datetime


class ExecutionEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    run_id: UUID
    sequence: int
    type: str
    timestamp: datetime
    node_id: str | None = None
    agent: str | None
    action: str
    status: str
    level: str
    duration_ms: int | None
    output_summary: str | None
    payload: dict[str, object]

    @classmethod
    def from_event(cls, event: "ExecutionEvent") -> "ExecutionEventResponse":
        return cls(
            id=event.id,
            run_id=event.run_id,
            sequence=event.sequence,
            type=event.event_type,
            timestamp=event.created_at,
            node_id=event.node_key,
            agent=event.agent,
            action=event.action,
            status=event.status,
            level=event.level,
            duration_ms=event.duration_ms,
            output_summary=event.output_summary,
            payload=event.payload,
        )


class ExecutionLogResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    node_key: str
    node_label: str
    status: str
    duration_ms: int
    output_summary: str | None
    error: str | None
    details: dict[str, object]
    created_at: datetime


class ExecutionDetailResponse(BaseModel):
    run: WorkflowRunResponse
    events: list[ExecutionEventResponse]
    logs: list[ExecutionLogResponse]


class RuntimeCommand(BaseModel):
    reason: str | None = Field(default=None, max_length=500)


class PaymentConfirmationCommand(BaseModel):
    node_id: str = Field(min_length=1, max_length=160)
    transaction_hash: str = Field(min_length=32, max_length=128)
